linkerd/error-respond/src/lib.rs

```diff
@@ -7,17 +7,20 @@ use linkerd2_error::Error;
 
 /// Creates an error responder for a request.
 pub trait NewRespond<Req, E = Error> {
-    type Response;
-    type Respond: Respond<E, Response = Self::Response>;
+    type ResponseIn;
+    type ResponseOut;
```

Comment:
@olix0r Maybe I am missing something here, but do you think there is a way to get rigt of the extra type here and still be able to map the response body only in the case when we are GRPC.

Reply:
I would probably do this as

```rust
pub trait NewRespond<Req, Rsp, E = Error> {
    type Response
    type Respond: Respond<Rsp, E, Response = Self::Response>;
```

The general rule being that the Req/Rsp/E are all passed into the the trait (and not returned by it). We don't want to have to have a concrete in-response type. Implementations should be generic over it.

Reply:
Okey, maybe I am way off here, but in essence what is happening here is that we get one response type and map it to another, right. The concrete example is an http:Response<SomePayload> -> http::Response<SomePayloadThatKnowsAboutGrpcErrors>. So don't we need two type params for that? Because at the end in the implementation in errors.rs this all maps to getting an http::Response<HttpBody> and mapping it to http::Response<errors::ResponseBody>.

Reply:
You are correct, except this crate is protocol-agnostic.  So the impl will only be generic over body like:

```rust
impl<ReqB, RspB, E> NewRespond<http::Request<ReqB>, http::Response<RspB>, E> for MyImpl<ReqB, RspB, E>
```

Reply:
Got it. Makes sense